boa/src/vm/opcode.rs

```diff
@@ -0,0 +1,614 @@
+    ///
+    /// Operands: name_index: `u32`
+    ///
+    /// Stack: **=>**
+    DefVar,
+
+    /// Declate `let` type variable.
+    ///
+    /// Operands: name_index: `u32`
+    ///
+    /// Stack: **=>**
+    DefLet,
+
+    /// Declate `const` type variable.
+    ///
+    /// Operands: name_index: `u32`
+    ///
+    /// Stack: **=>**
+    DefConst,
+
+    /// Initialize a lexical binding.
+    ///
+    /// Operands: name_index: `u32`
+    ///
+    /// Stack: **=>**
+    InitLexical,
+
+    /// Find a binding on the environment chain and push its value.
+    ///
+    /// Operands: name_index: `u32`
+    ///
+    /// Stack: **=>** value
+    GetName,
+
+    /// Find a binding on the environment chain and assign its value.
+    ///
+    /// Operands: name_index: `u32`
+    ///
+    /// Stack: value **=>**
+    SetName,
+
+    /// Get a property by name from an object an push it on the stack.
+    ///
+    /// Like `object.name`
+    ///
+    /// Operands: name_index: `u32`
+    ///
+    /// Stack: object **=>** value
+    GetPropertyByName,
+
+    /// Get a property by value from an object an push it on the stack.
+    ///
+    /// Like `object[key]`
+    ///
+    /// Operands:
+    ///
+    /// Stack: key, object **=>** value
+    GetPropertyByValue,
+
+    /// Sets a property by name of an object.
+    ///
+    /// Like `object.name = value`
+    ///
+    /// Operands: name_index: `u32`
+    ///
+    /// Stack: value, object **=>**
+    SetPropertyByName,
+
+    /// Sets a property by value of an object.
+    ///
+    /// Like `object[key] = value`
+    ///
+    /// Operands:
+    ///
+    /// Stack: value, key, object **=>**
+    SetPropertyByValue,
+
+    /// Unconditional jump to address.
+    ///
+    /// Operands: address: `u32`
+    /// Stack: **=>**
+    Jump,
+
+    /// Constional jump to address.
+    ///
+    /// If the value popped is [`falsy`][falsy] then jump to `address`.
+    ///
+    /// Operands: address: `u32`
+    ///
+    /// Stack: cond **=>**
+    ///
+    /// [falsy]: https://developer.mozilla.org/en-US/docs/Glossary/Falsy
+    JumpIfFalse,
+
+    /// Constional jump to address.
+    ///
+    /// If the value popped is [`truthy`][truthy] then jump to `address`.
+    ///
+    /// Operands: address: `u32`
+    ///
+    /// Stack: cond **=>**
+    ///
+    /// [truthy]: https://developer.mozilla.org/en-US/docs/Glossary/Truthy
+    JumpIfTrue,
+
+    /// Throw exception
+    ///
+    /// Operands:
+    ///
+    /// Stack: `exc` **=>**
+    Throw,
+
+    /// Pops value converts it to boolean and pushes it back.
+    ///
+    /// Operands:
+    ///
+    /// Stack: value **=>** (`ToBoolean(value)`)
+    ToBoolean,
+
+    /// Pushes `this` value
+    ///
+    /// Operands:
+    ///
+    /// Stack: **=>** `this`
+    This,
+
+    /// No-operation instruction, does nothing.
+    ///
+    /// Operands:
+    ///
+    /// Stack: **=>**
+    // Safety: Must be last in the list since, we use this for range checking
```

Comment:
I understand this, but if we ever add more opcodes, it would make it backwards incompatible.

Maybe it would be better to assign a number to each and have a `TryFrom` implementation that checks a hardcoded number.

Reply:
> I understand this, but if we ever add more opcodes, it would make it backwards incompatible.

Could you explain what you meant by this? We can add opcodes wherever we want as long as `Nop` is the last. we can even shuffle then and it will still work.

> Maybe it would be better to assign a number to each and have a `TryFrom` implementation that checks a hardcoded number.

But if we did this we would have to maintain the current opcode number (like Pop is `0`, Push is `1` etc) and adding opcodes we would have to increase this number manually, and shufulling them would be very difficult since it would require updating all the preceding hardcoded opcode values. (we may want to do this to group some opcodes by behaviour like `GetPropertyByName`, `GetPropertyByValue`, `SetPropertyByName` `SetPropertyByValue` that access properties of objects).

> Another option would be to hardcode one number per opcode and then have a constant that would be the "maximum opcode". We would need a test to make sure that we update it if we ever add an opcode.

Even with this it requires some maintenance cost like the previous solution.

Reply:
You mean something like [v8 snapshot](https://github.com/danbev/learning-v8/blob/master/notes/snapshots.md)? that we can restore later, right?

I don't think they could be compatible between versions since we can ass more opcodes that the previous versions would not understand.

Reply:
Created issue for start up snapshots #1372

Reply:
> You mean something like [v8 snapshot](https://github.com/danbev/learning-v8/blob/master/notes/snapshots.md)? that we can restore later, right?

Actually, I was thinking more of a Python cache. If you are running the same JS script multiple times, you can parse and compile it once, save the instructions to a file, and just execute the output instead of re-compiling every time.